src/StaticLS/IDE/CodeActions.hs

```diff
@@ -5,32 +5,74 @@
 
 module StaticLS.IDE.CodeActions where
 
+import Control.Error (mapMaybe)
+import Control.Monad (join)
 import Data.LineCol (LineCol (..))
 import Data.Path (AbsPath)
 import Data.Rope qualified as Rope
+import Data.Set qualified as S
+import Language.LSP.Protocol.Types qualified as LSP
+import StaticLS.IDE.CodeActions.AddRequiredExtension qualified as AddRequiredExtension
 import StaticLS.IDE.CodeActions.AddTypeSig qualified as AddTypeSig
 import StaticLS.IDE.CodeActions.AutoImport qualified as AutoImport
-import StaticLS.IDE.CodeActions.RemoveRedundantImports as RemoveRedundantImports
+import StaticLS.IDE.CodeActions.InsertAssociatedType qualified as InsertAssociatedType
+import StaticLS.IDE.CodeActions.InsertCases qualified as InsertCases
+import StaticLS.IDE.CodeActions.InsertFields qualified as InsertFields
+import StaticLS.IDE.CodeActions.InsertMissingMethods qualified as InsertMissingMethods
+import StaticLS.IDE.CodeActions.Parse qualified as Parse
+import StaticLS.IDE.CodeActions.RemoveRedundantImports qualified as RemoveRedundantImports
 import StaticLS.IDE.CodeActions.Types
+import StaticLS.IDE.CodeActions.UseValidHoleFit qualified as UseValidHoleFit
 import StaticLS.IDE.Monad
 import StaticLS.IDE.SourceEdit (SourceEdit)
 import StaticLS.IDE.SourceEdit qualified as SourceEdit
 import StaticLS.Monad (StaticLsM)
+import StaticLS.ProtoLSP (assistToCodeAction)
 
-getCodeActions :: AbsPath -> LineCol -> StaticLsM [Assist]
-getCodeActions path lineCol = do
+getCodeActions ::
+  LSP.TextDocumentIdentifier ->
+  LSP.CodeActionContext ->
+  AbsPath ->
+  LineCol ->
+  StaticLsM [LSP.CodeAction]
+getCodeActions tdi ctx path lineCol = do
   rope <- getSourceRope path
   let pos = Rope.lineColToPos rope lineCol
   let cx = CodeActionContext {path, pos, lineCol}
   typesCodeActions <- AddTypeSig.codeAction cx
   importCodeActions <- AutoImport.codeAction cx
   removeRedundantImports <- RemoveRedundantImports.codeAction cx
-  let codeActions = typesCodeActions ++ importCodeActions ++ removeRedundantImports
-  pure codeActions
+  let issues = S.fromList (mapMaybe Parse.actionableIssue ctx._diagnostics)
+  issueActions <- join <$> traverse (issueToActions tdi) (S.toList issues)
+  assistActions <-
+    traverse assistToCodeAction $
```

Comment:
See [discussion above](https://github.com/josephsumabat/static-ls/pull/164/files#r2188513659) for rationale on moving `assistToCodeAction` into the body of `codeAction`.